### PR TITLE
Remove transpilation of Ably React Hooks and re-implement history via SSR (and ISR)

### PR DESCRIPTION
**Pull request by marklewin**
@leggetter I was hoping you could have a look at this before I adjust the blog post as per your request.

Items addressed:

(1) Removed the whole `next-transpile-modules` workflow for Ably React Hooks. Jo Franchetti has updated the React Hooks package so that, as of v2.0.4, this is no longer required.

(2) Re-implemented History. For context, this is why Jo removed it:

> Using rewind instead of the history API, fewer roundtrips to the server, also ensures correct message order.

and

> deleted because if you are connecting to a channel and need to get its history, you should use rewind to prevent race conditions where current messages arrive on the channel as history is being retrieved, which could lead to message ordering issues during render.

I haven't noticed any issues with either of the above. The way is works is that, when the `index.js` page is loaded, there is a call to `getStaticProps` to retrieve the last five items from history and then any further updates are retrieved in realtime. I can specify [ISR (incremetal static regeneration)](https://nextjs.org/docs/basic-features/data-fetching/incremental-static-regeneration) with the `revalidate` property in that function (see code comment), but I don't really see the point. There is only ever a single call to history when the page first loads and all newer articles appear in realtime.

I'd like to know your thoughts on this before I rewrite the affected portions of the blog post? Thanks :)

### CODE REVIEW

**Comment by leggetter** [comment]
ISR is beneficial because you've created a static site but you need some content to be statically regenerated every so often. You don't want to regenerate the whole site, you just want certain pages to auto-regenerate. Revalidate lets you configure the period of auto-revalidation.

The main thing we're looking to demonstrate is that you can use Ably as a form of cache so we want some content in the page at static generation time. There are SEO benefits to this.

The points around ordering and race conditions are important. I wonder, therefore, whether "best practice" would be to use both the server-side things for static generation/SEO but to replace the content on the client for those guarantees? Or can we pre-load a last message ID from the server and then use that ID with the realtime/client-side subscription to ensure we don't have a race condition? I can't find an obvious way of doing this 😊 If it's not possible and we have concerns then maybe replacing the content on the client is the only (and simplest) approach?

File: components/Articles.js [hunk 1]
```diff
@@ -3,13 +3,13 @@ import { useChannel } from "@ably-labs/react-hooks";
...
-export default function Articles() {
+export default function Articles(props) {
   let inputBox = null;
 
   const [headlineText, setHeadlineText] = useState("");
-  const [headlines, updateHeadlines] = useState([]);
-  const [_, ably] = useChannel("[?rewind=3]headlines", (headline) => {
-    updateHeadlines((prev) => [...prev, headline]);
+  const [headlines, updateHeadlines] = useState(props.history);
+  const [_, ably] = useChannel("headlines", (headline) => {
```
Comment (leggetter) [inline components/Articles.js]: See top-level comments about ordering etc. If we have concerns about this I'd also keep the history subscription (of 5 to be consistent with the SSR?) here to ensure we don't miss messages during the load time.

File: lib/history.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+
+const rest = new Ably.Rest.Promise({
+  key: process.env.ABLY_SERVER_API_KEY,
+});
+
+const channel = rest.channels.get("headlines");
+
+export async function getHistoricalMessages() {
+  const resultPage = await channel.history({ limit: 5 });
+  const historicalMessages = JSON.parse(JSON.stringify(resultPage.items));
```
Comment (leggetter) [inline lib/history.js:15]: Why do we need to do this?

Stringify turns into a string and then we parse from a strong back to JSON.

Comment (marklewin) [reply lib/history.js:15]: Yes, this is a very annoying [issue](https://github.com/vercel/next.js/issues/11993). It seems that you have to re-parse or certain Message object items can't be serialized by props later.

**Comment by leggetter** [comment]
@marklewin and I had a video call.

The plan:

- [ ] Bring the client-side history functionality back in to this PR to deal with the race condition of a message coming in during the browser fetching and loading the page and Ably connection events.
- Add details of the ISR to the blog post explaining that it provides the benefits of:
  - [ ] SEO through the statically generated asset containing some real headline content
  - [ ] The page loads with populated data and there isn't a pop effect as data is loaded into an initially empty area after the Ably connection is established and data is received.
- [ ] Update the blog post to talk about the race condition to the blog post regarding the `rewind` requirement

File: lib/history.js (hunk 2 above)
Comment (leggetter) [reply lib/history.js:15]: > Yes, this is a very annoying [issue](https://github.com/vercel/next.js/issues/11993). It seems that you have to re-parse or certain Message object items can't be serialized by props later.

😱

**Comment by marklewin** [comment]
@leggetter I hope I have understood this correctly. What I've tried to achieve with the latest commit is the following:
- Render static history with ISR enabled when the page first loads
- When the client connects to the channel, replace the history entries with those from channel.rewind

I'll hold off updating any blog changes until I'm clear that this is what is expected. Thanks.

File: components/Articles.js [hunk 3]
```diff
@@ -3,15 +3,25 @@ import { useChannel } from "@ably-labs/react-hooks";
 import ArticlePreview from "./ArticlePreview";
 import styles from "../styles/Home.module.css";
 
-export default function Articles() {
+let clearHistoryState = true;
```
Comment (leggetter) [inline components/Articles.js:12]: Add a comment to this variable to explain why it's here. The variable name is good but I do think this is one of those cases where a comment would help clarify.

File: lib/history.js (hunk 2 above)
Comment (leggetter) [reply lib/history.js:15]: Recommend: Add comment about `JSON.parse(JSON.stringify(` and link to the issue.

**Review by leggetter — APPROVED** [review]

File: components/Articles.js (hunk 1 above)
Comment (marklewin) [reply components/Articles.js]: Done in [a65cd39](https://github.com/ably-labs/ably-next-vercel-news/pull/8/commits/a65cd397ed945fe13c86d64b8f36fffaa563f98b)

File: lib/history.js (hunk 2 above)
Comment (marklewin) [reply lib/history.js:15]: Done in [a65cd39](https://github.com/ably-labs/ably-next-vercel-news/pull/8/commits/a65cd397ed945fe13c86d64b8f36fffaa563f98b)

**Pull request merged**